### ISSUE TITLE
Nondeterministic Crash in Dtype.TryGetTypeCore

### ISSUE THREAD

**Issue by MikeStall** [microsoft/Power-Fx#1511]
Got crash at this callstack. Succeeded on retry: 
![image](https://github.com/microsoft/Power-Fx/assets/1538900/fb684a9c-b04a-4815-9a28-a15c8af7aece)

**Comment by MikeStall** [comment]
I can repro locally with this test:
```
        public class MyType2
        {
            public string Field1 { get; set; }
        }

        [Fact]
        public void MarshalStress()
        {
            var engine = new Engine();

            for (int i = 0; i < 1000; i++)
            {
                var cache = new TypeMarshallerCache();
                RecordType recordType = (RecordType)cache.GetMarshaller(typeof(MyType2)).Type;

                Parallel.For(
                    0,
                    3,
                    (j) =>
                    {
                        var i2 = i;
                        var symbolTable = new SymbolTable();
                        symbolTable.AddVariable("record", recordType);

                        var check = engine.Check("record.Field1", symbolTable: symbolTable);
                        Assert.True(check.IsSuccess);
                    });
            }
        }
```

**Comment by MikeStall** [comment]
Fixed with #1516
